topdown/reachable.go

```diff
@@ -0,0 +1,54 @@
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func builtinReachable(graphVal, initialVal ast.Value) (ast.Value, error) {
+	graph, err := builtins.ObjectOperand(graphVal, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	initial, err := builtins.SetOperand(initialVal, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// This is a queue that holds all nodes we still need to visit.  It is
+	// initialised to the initial set of nodes we start out with.
+	queue := []*ast.Term{}
+	initial.Foreach(func(t *ast.Term) {
+		queue = append(queue, t)
+	})
+
+	// This is the set of nodes we have reached.
+	reached := ast.NewSet()
+
+	// Keep going as long as we have nodes in the queue.
+	for len(queue) > 0 {
+		// Get the edges for this node.  If the node was not in the graph,
+		// `edgesTerm` will be `nil` and we can ignore it.
+		node := queue[0]
+		if edgesTerm := graph.Get(node); edgesTerm != nil {
+			// Add all the newly discovered neighbors.
+			if edgeSet, ok := edgesTerm.Value.(ast.Set); ok {
+				edgeSet.Foreach(func(neighbor *ast.Term) {
+					if !reached.Contains(neighbor) {
+						queue = append(queue, neighbor)
+					}
+				})
+			}
+			// Mark the node as reached.
+			reached.Add(node)
+		}
+		queue = queue[1:]
+	}
+
+	return reached, nil
+}
+
+func init() {
+	RegisterFunctionalBuiltin2(ast.Reachable.Name, builtinReachable)
```

Comment:
nit: Use `RegisterBuiltinFunc` https://github.com/open-policy-agent/opa/blob/7fdb6a178042d4f0bdba0a9deb4d54316a034601/topdown/builtins.go#L52 as the other ones have been deprecated.